cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -90,181 +129,223 @@ func newCmdApply(apf *applyPlanFlags) *cobra.Command {
 	cmd.Flags().BoolVar(&flags.renewCerts, options.CertificateRenewal, flags.renewCerts, "Perform the renewal of certificates used by component changed during upgrades.")
 	options.AddPatchesFlag(cmd.Flags(), &flags.patchesDir)
 
+	// Initialize the workflow runner with the list of phases
+	applyRunner.AppendPhase(phases.NewPreflightPhase())
+	applyRunner.AppendPhase(phases.NewControlPlanePhase())
+	applyRunner.AppendPhase(phases.NewUploadConfigPhase())
+	applyRunner.AppendPhase(phases.NewKubeconfigPhase())
+	applyRunner.AppendPhase(phases.NewKubeletConfigPhase())
+	applyRunner.AppendPhase(phases.NewBootstrapTokenPhase())
+	applyRunner.AppendPhase(phases.NewAddonPhase())
```

Comment:
we might consider adding a phase now or later at the end here. it can be for the post upgrade tasks like updating the /var/lib/kubelet/kubeadm-flags.env or patching some config map.

Reply:
Added a new empty `post-upgrade` phase

Reply:
@neolit123 Thanks for the detailed review, the code has been updated. PTAL again.

Reply:
i think we might want to add the same `post-upgrade` phase for `upgrade node`?
we should add this to the TODO list that must be done after this PR merges.